tagilmo/utils/malmo_wrapper.py

```diff
@@ -280,13 +289,19 @@ def getCachedObserve(self, method, key = None):
     def observeProcCached(self):
         self.mc.observeProc()
         t_new = time.time()
+        updated = False
         for method in self.methods:
             v_new = getattr(self.mc, method)(self.nAgent)
             v, t = self.cached[method]
             outdated = t_new - t > self.max_dt
             if v_new is not None or outdated: # or v is None
+                self.cached_buffer[method] = self.cached[method]
                 self.cached[method] = (v_new, t_new)
                 self.changed(method)
+                updated = True
+        if updated:
+            self.cached_buffer_list.append(self.cached_buffer.copy())
+            self.cached_buffer_list = self.cached_buffer_list[-self.cbuff_history_len:]
```

Comment:
you can use deque - it can automatically remove old item when hitting length limit

Reply:
Yeah, I thought about it. As far as I remember you can not do stuff like a[:2] with deque. That's why I preferred vanilla lists for small buffer size. For the bigger once there is obvious  advantage for deque.